store/mongo/datastore_mongo.go

```diff
@@ -368,22 +375,47 @@ func (db *DataStoreMongo) upsertAttributes(
 	case 1:
 		var res *mongo.UpdateResult
 		if withRevision {
-			filter = bson.M{
-				"_id":         devices[0].Id,
-				DbDevRevision: bson.M{"$lt": devices[0].Revision},
+			if etag != nil && *etag != "" {
+				filter = bson.M{
+					"_id":         devices[0].Id,
+					DbDevRevision: bson.M{"$lt": devices[0].Revision},
+					etagField:     bson.M{"$exists": true, "$eq": *etag},
+				}
+				if err := checkMatchedCount(ctx, c, filter); err != nil {
+					return &model.UpdateResult{MatchedCount: 0}, nil
+				}
+
+			} else {
+				filter = bson.M{
+					"_id":         devices[0].Id,
+					DbDevRevision: bson.M{"$lt": devices[0].Revision},
+				}
 			}
 			update[DbDevRevision] = devices[0].Revision
-			update = bson.M{
-				"$set":         update,
-				"$setOnInsert": oninsert,
-			}
 		} else {
-			filter = map[string]interface{}{"_id": devices[0].Id}
-			update = bson.M{
-				"$set":         update,
-				"$setOnInsert": oninsert,
+			if etag != nil && *etag != "" {
+				filter = bson.M{
+					"_id":     devices[0].Id,
+					etagField: bson.M{"$exists": true, "$eq": *etag},
+				}
+				// check if we are going to update existing document with the
```

Comment:
I think, even with this check, it will still be possible to get duplicate key error under some conditions (with concurrent updates) and it looks like we have a handling for this kind of errors, but for a different case, which makes it hard to distinguish both cases (maybe using a scope to distinguish them would be an option?)

Reply:
Great idea! Thanks!

Reply:
Initially I thought that building logic based on errors is not the best, but in this case it looks way better than extra DB call + not handled cases.